File: 005/cpp/main.cpp
class Solution {
public:
    string longestPalindrome1(string s, int i) {
        int n = std::min<int>(i, s.size() - i - 1);
        int j = 1;
        for (; j < n + 1; j++) {
            if (s[i - j] != s[i + j]) {
                break;
            }
        }
        return s.substr(i - j + 1, 2 * j - 1);
    }

    string longestPalindrome2(string s, int i) {
        int n = std::min<int>(i + 1, s.size() - i - 1);
        int j = 1;
        for (; j < n + 1; j++) {
            if (s[i - j + 1] != s[i + j]) {
                break;
            }
        }
        return s.substr(i - j + 2, 2 * j - 2);
    }

    string longestPalindrome(string s) {
        string r = "";
        for (int i = 0; i < s.size(); i++) {
            string r0;
            r0 = longestPalindrome1(s, i);
            if (r0.size() > r.size()) {
                r = r0;
            }
            r0 = longestPalindrome2(s, i);
            if (r0.size() > r.size()) {
                r = r0;
            }
        }
        return r;
    }
};

File: 005/rust/main.rs
impl Solution {
    fn longest_palindrome1(s: &String, i: usize) -> String {
        let s0 = s.as_bytes();
        let n = std::cmp::min(i, s0.len() - i - 1);
        let mut k = 0;
        for j in 1..n + 1 {
            k = j;
            if (s0[i - j] != s0[i + j]) {
                k = j - 1;
                break;
            }
        }
        return s[i - k..i + k + 1].to_string();
    }

    fn longest_palindrome2(s: &String, i: usize) -> String {
        let s0 = s.as_bytes();
        let n = std::cmp::min(i + 1, s0.len() - i - 1);
        let mut k = 0;
        for j in 1..n + 1 {
            k = j;
            if (s0[i - j + 1] != s0[i + j]) {
                k = j - 1;
                break;
            }
        }
        if k == 0 {
            return "".to_string();
        }
        return s[i - k + 1..i + k + 1].to_string();
    }

    pub fn longest_palindrome(s: String) -> String {
        let mut r = "".to_string();
        for i in 0..s.len() {
            let mut r0;
            r0 = Solution::longest_palindrome1(&s, i);
            if (r0.len() > r.len()) {
                r = r0;
            }
            r0 = Solution::longest_palindrome2(&s, i);
            if (r0.len() > r.len()) {
                r = r0;
            }
        }
        return r;
    }
}
